### PR TITLE
Codegen

### PR DESCRIPTION
**Pull request by sahejsingh**
### Code generator for Sysl

Use sysl transforms and ebnf grammar of the target language to generate code for sysl model.

@anz-bank/sysl-developers

### CODE REVIEW

File: sysl2/sysl/codegen.go [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+	root := processChoice(g, obj, rule.Choices)
+	if root == nil {
+		logrus.Printf("could not process rule: ( %s )", ruleName)
+	}
+	logrus.Printf("Processed rule: ( %s ), len(%d)", ruleName, len(root))
+	return root
+}
+
+func readGrammar(filename, grammarName, startRule string) *ebnfGrammar.Grammar {
+	dat, _ := ioutil.ReadFile(filename)
```
Comment (anzdaddy) [inline sysl2/sysl/codegen.go]: Check error

File: sysl2/sysl/codegen.go [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+	}
+}
+
+// return the one and only app defined in the module
+func getDefaultAppName(mod *sysl.Module) string {
+	var apps []string
+	for app := range mod.Apps {
+		apps = append(apps, app)
+	}
+	return apps[0]
```
Comment (anzdaddy) [inline sysl2/sysl/codegen.go]: ??

File: sysl2/sysl/codegen.go [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+// GenerateCode transform input sysl model to code in the target language described by
+// grammar and a sysl transform
+func GenerateCode(root_model, model, root_transform, transform, grammar, start string) []*codeGenOutput {
+	var codeOutput []*codeGenOutput
+
+	mod, modelAppName := loadAndGetDefaultApp(root_model, model)
+	tx, transformAppName := loadAndGetDefaultApp(root_transform, transform)
+	g := readGrammar(grammar, "gen", start)
+	if g == nil {
+		panic("unable to parse grammar")
```
Comment (anzdaddy) [inline sysl2/sysl/codegen.go]: Return `error`.

File: sysl2/sysl/examples/grammars/go/go.gen.sysl [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      let isKeyword = IsKeyword(field.key).out
+      let prefix = if isKeyword == true then "_" else ""
+      identifier = prefix + field.key
+      let IsPtr = if field.value.optional == true then "*" else ""
+      Type = IsPtr + GoType(field.value).out
+    )
+
+  !view GoStructs(types <: set of Type) -> set of TopLevelDecl:
+    types -> (type:
+        Declaration = type -> <Declaration>(:
```
Comment (anzdaddy) [inline sysl2/sysl/examples/grammars/go/go.gen.sysl]: Is `<Declaration>` omittable, given that the interpreter is dynamically typed, anyway?

File: sysl2/naive/Grammar1_test.go [hunk 5]
```diff
@@ -1,30 +1,34 @@
...
+package parser
 
 import (
 	"testing"
 
 	sysl "github.com/anz-bank/sysl/sysl2/proto"
 )
 
-func testParser(g *sysl.Grammar, numTerms int, tokens []int, text string, expected bool, t *testing.T) {
+func testParser(g *sysl.Grammar, numTerms int, tokens []int, text string, expectedResult bool, t *testing.T) (bool, []interface{}) {
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/naive/Grammar1_test.go]: Not necessarily. testParser is not "THE" Test function.

[the remark this answers is not in the document]
Comment (anzdaddy) [reply sysl2/naive/Grammar1_test.go]: The reason is that `t` is essentially the function's operating context, not unlike `context.Context`. It's not a big deal.

File: sysl2/naive/ebnfParser.go [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+		panic("not implemented yet.")
+	}
+	return a
+}
+
+func symbolTerm(item interface{}) symbol {
+	return item.(symbol)
+}
+
+func getChoice(choice map[int][]interface{}) (int, []interface{}) {
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/naive/ebnfParser.go]: no. it can be different. Only "choice" rule has value 0.

File: sysl2/sysl/ValueUtil.go [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+		if typeName == "sequence" {
+			seqMap := MakeValueMap()
+			addItemToValueMap(m.GetMap(), typeName, seqMap)
+
+			typeName, typeDetail = getTypeDetail(t.GetSequence())
+			addItemToValueMap(seqMap.GetMap(), "type", MakeValueString(typeName))
+			addItemToValueMap(seqMap.GetMap(), typeName, MakeValueString(typeDetail))
+		}
+
+		addItemToValueMap(fieldMap.GetMap(), key, m)
```
Comment (anzdaddy) [inline sysl2/sysl/ValueUtil.go]: This is confusing right in the middle of populating `m`. Move it to the end.

File: sysl2/sysl/ValueUtil.go [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+// AddApp add sysl.App to scope
+func (s *Scope) AddApp(name string, app *sysl.Application) {
+	m := MakeValueMap()
+	(*s)[name] = m
+	addItemToValueMap(m.GetMap(), "name", MakeValueString(getAppName(app.Name)))
+	addItemToValueMap(m.GetMap(), "attrs", attrsToValueMap(app.Attrs))
+	addItemToValueMap(m.GetMap(), "types", typesToValueMap(app.Types))
+	addItemToValueMap(m.GetMap(), "union", unionToValueMap(app.Types))
```
Comment (anzdaddy) [inline sysl2/sysl/ValueUtil.go]: This is a good example of where a parameter to `MakeValueMap()` would really clean things up:
```go
	(*s)[name] = MakeValueMap({
		"name": MakeValueString(getAppName(app.Name)),
		"attrs": attrsToValueMap(app.Attrs),
		"types": typesToValueMap(app.Types),
		"union": unionToValueMap(app.Types),
	})
```

File: sysl2/sysl/exprEval.go [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+		case *sysl.Expr_Transform_Stmt_Let:
+			res := Eval(txApp, assign, ss.Let.Expr)
+			(*assign)[ss.Let.Name] = res
+			logrus.Printf("Eval %s: %v\n", ss.Let.Name, res)
+		case *sysl.Expr_Transform_Stmt_Assign_:
+			logrus.Printf("Eval %s:\n", ss.Assign.Name)
+			res := Eval(txApp, assign, ss.Assign.Expr)
+			logrus.Printf("Eval %s ==\n\t\t %v:\n", ss.Assign.Name, res)
+			addItemToValueMap(result.GetMap(), ss.Assign.Name, res)
+			// case *sysl.Expr_Transform_Stmt_Inject:
```
Comment (anzdaddy) [inline sysl2/sysl/exprEval.go]: Remove comment or add TODO.

File: sysl2/sysl/exprEval.go [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+	case *sysl.Expr_Set:
+		{
+			setResult := MakeValueSet()
+			v := []*sysl.Value{}
+			for _, s := range x.Set.Expr {
+				v = append(v, Eval(txApp, assign, s))
+			}
+			setResult.GetSet().Value = v
+			return setResult
+		}
```
Comment (anzdaddy) [inline sysl2/sysl/exprEval.go]: Another example that would read more easily with an initializer for `MakeValueSet()`:
```go
case *sysl.Expr_Set:
	{
		v := []*sysl.Value{}
		for _, s := range x.Set.Expr {
			v = append(v, Eval(txApp, assign, s))
		}
		return MakeValueSet(v)
	}
```

File: sysl2/sysl/tests/eval_expr.sysl [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    lhs -> (:
+      out1 = lhs + rhs
+      out2 = lhs -> <int> (:
+        out3 = 5 + 1
+      )
+    )
+
+  !view addSet(lhs <: string) -> set of string:
+    lhs -> (:
+      out = {"lhs"} | {"rhs"}
```
Comment (anzdaddy) [inline sysl2/sysl/tests/eval_expr.sysl]: `{"lhs", "rhs"}`

File: sysl2/naive/ebnfParser.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/naive/ebnfParser.go]: choice is a map, there is only one key, but that key can be any `int`

File: sysl2/naive/grammarParser.go [hunk 12]
```diff
@@ -79,8 +91,143 @@ func makeParser(g *sysl.Grammar, text string) *parser {
...
+		tree = append(tree, nil)
+	}
+	return result, input, tree
+}
+
+// parseGrammar returns true if the grammar consumes the whole string
+// g - grammar to use
+// text - to parse
+// Start rule
+func (p *parser) parseGrammar(arr *[]token) (bool, []interface{}) {
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/naive/grammarParser.go]: bool tells us if the parse was successful. if true, look at the second arg (ast)

File: sysl2/naive/Grammar1_test.go [hunk 13]
```diff
@@ -251,3 +248,116 @@ func TestFirstSet2(t *testing.T) {
...
+	grammars := []string{
+		`s : 'd' | 'c' s ; `,
+		`s : 'c'* 'd' ; `,
+	}
+	text := "ccd"
+	tokens := [][]int{
+		{1, 1, 0, -1},
+		{0, 0, 1, -1},
+	}
+	choiceBranch := []int{1, 0}
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/naive/Grammar1_test.go]: done.

File: sysl2/naive/ebnfParser.go [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+
+func buildChoice(choice []interface{}) *sysl.Choice {
+	choiceS := make([]*sysl.Sequence, 0)
+	var s0 []interface{}
+	_, s0 = ruleSeq(choice[0], "seq")
+
+	choiceS = append(choiceS, buildSequence(s0))
+	if len(choice) > 1 {
+		x := choice[1].([]interface{})
+		if x[0] != nil {
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/naive/ebnfParser.go]: []interface{}[0], need a magnifying glass to read that :(

File: sysl2/sysl/ValueUtil.go [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+func MakeValueString(val string) *sysl.Value {
+	return &sysl.Value{
+		Value: &sysl.Value_S{
+			S: val,
+		},
+	}
+}
+
+// MakeValueMap ...
+func MakeValueMap() *sysl.Value {
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/sysl/ValueUtil.go]: I am keeping the Make functions for now. need append.. functions in quite a few places.

File: sysl2/sysl/examples/grammars/go/go.gen.sysl (hunk 4 above)
Comment (sahejsingh) [reply sysl2/sysl/examples/grammars/go/go.gen.sysl]: gosysl checks expects a transform to return a type.

File: sysl2/naive/Grammar1_test.go [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+	g := makeEXPR()
+	terms := makeBuilder().buildTerminalsList(g.Rules)
+
+	if len(terms) != 7 {
+		t.Error("got incorrect number of terms", terms)
+	}
+
+	first, follow := buildFirstFollowSet(g)
+
+	if len(first) < 0 || len(follow) < 0 {
```
Comment (anzdaddy) [inline sysl2/naive/Grammar1_test.go]: This can't fail.

File: sysl2/naive/Grammar1_test.go [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+	g := makeG2()
+	terms := makeBuilder().buildTerminalsList(g.Rules)
+
+	if len(terms) != 3 {
+		t.Error("got incorrect number of terms", terms)
+	}
+
+	first, follow := buildFirstFollowSet(g)
+
+	if len(first) < 0 || len(follow) < 0 {
```
Comment (anzdaddy) [inline sysl2/naive/Grammar1_test.go]: This can't fail.

**Pull request merged**